dev/io.openliberty.security.jakartasec.4.0.internal/src/io/openliberty/security/jakartasec40/identitystore/InMemoryIdentityStore.java

```
@@ -0,0 +1,161 @@
+        inMemoryIdentityStoreDefinitionWrapper = new InMemoryIdentityStoreDefinitionWrapper(inMemoryIdentityStoreDefinition);
+    }
+
+    @Override
+    public Set<String> getCallerGroups(CredentialValidationResult validationResult) {
+        Set<String> groups = new HashSet<String>();
+        if (!validationTypes().contains(ValidationType.PROVIDE_GROUPS)) {
+            return groups; // empty
+        }
+
+        String caller = validationResult.getCallerPrincipal().getName();
+        if (caller == null) {
+            if (tc.isEventEnabled()) {
+                Tr.event(tc, "A null caller was passed into getCallerGroups. No groups returned. " + validationResult);
+            }
+            return groups;
+        }
+
+        CredentialValue credentialValue = inMemoryIdentityStoreDefinitionWrapper.getCredential(caller);
+        if (credentialValue != null) {
+            // found a caller, so set groups
+            groups = Set.of(credentialValue.getGroups());
+        }
+
+        return groups;
+    }
+
+    @Override
+    public int priority() {
+        return inMemoryIdentityStoreDefinitionWrapper.getPriority();
+    }
+
+    @Override
+    @Sensitive
+    public CredentialValidationResult validate(Credential credential) {
+
+        // first check the annotation allows for validation
+        if (!validationTypes().contains(ValidationType.VALIDATE)) {
+            return CredentialValidationResult.NOT_VALIDATED_RESULT;
+        }
+
+        // now extract client caller and password details
+        boolean callerOnly = false;
+        String caller;
+        ProtectedString password = null;
+        if (credential instanceof UsernamePasswordCredential) {
+            caller = ((UsernamePasswordCredential) credential).getCaller();
+            password = new ProtectedString(((UsernamePasswordCredential) credential).getPassword().getValue());
+        } else if (credential instanceof CallerOnlyCredential) {
```

In this case is it always going to fail validation?  Looking at CredentialValue.validate() it would seems so because any null or empty string passed in returns false.  So what is the point of checking CallerOnlyCredential?

---

Okay, more code cut and paste from Database/Ldap Identity Stores (where legitimate code flow conditions on `CallerOnlyCredential` do exist).  I've slightly restructured the code to fail faster with the same `INVALID_RESULT` result.

Done.